### PR TITLE
Make `StepPersistence` replay-safe under durable execution

### PR DESCRIPTION
**Pull request by DouweM**
Spun off from #710, which limited itself to four replay-safe capabilities and left `StepPersistence` out because it needs run-id and idempotency work rather than a plain adoption.

## What was broken

**The run id was random.** `_derive_run_id` was:

```python
if self.agent_name is not None:
    return f'{self.agent_name}-{uuid4().hex[:8]}'
return ctx.run_id or str(uuid4())
```

`uuid4()` is fresh on every call, so a recovered run derived a *different* run id and wrote a second, disconnected set of events and snapshots. That id is what every later write keys on, so nothing lined up after recovery: the point of the capability, resuming a run, was exactly what durable recovery could not do.

It now derives from `ctx.run_id`, using the **whole** value rather than a truncation. A caller can pass their own `run_id` to `Agent.run`, so it is not guaranteed to be a UUID and the last eight characters of an arbitrary string are neither well distributed nor collision-resistant. The `or str(uuid4())` fallbacks are gone: the graph assigns run ids before `for_run`, so `None` means something is wrong upstream, and inventing one there produces a silently unrecoverable run. `_effective_run_id` gets the same treatment.

**Timestamps were read in orchestration code.** `StepEvent.timestamp` and `ContinuableSnapshot.timestamp` both defaulted to `_utcnow()` at construction, and the records were built before being handed to the store. Under a durable engine that reads the wall clock outside the durable unit, so a replayed write recorded a different time, and on Temporal that clock read is sandbox-restricted in workflow code to begin with. Both are now created inside their durable operations.

**None of the writes were journaled**, so recovery re-entered the store.

## What this does

Five durable operations: run registration, event append, snapshot save, tool-effect start, and tool-effect completion. Plus a stable default `id='step_persistence'`, using the `KW_ONLY` idiom shipped for `SummarizingCompaction` in #707. That default is not cosmetic: `collect_capability_operations` is a class-level scan, so one `@durable_operation` method makes every instance contribute durable operations and durability binding rejects `id=None`. Without it this change would break every existing `StepPersistence` user who attaches a durability capability.

## Idempotency keys, and two design corrections worth recording

`ToolEffectRecord` already carried an `idempotency_key`. `StepEvent` and `ContinuableSnapshot` now do too, and both `InMemoryStepStore` and the Mongo store suppress a key they have already applied. A `None` key preserves today's append behaviour exactly, so a caller constructing records directly is unaffected.

Snapshots took two attempts to get right, and both wrong answers are worth naming because they look reasonable:

1. **Upserting on `(run_id, step_index)` is wrong.** A settled `complete` snapshot and a later `interrupted` rescue snapshot legitimately share a step index, and collapsing them destroys the safe snapshot so `latest_snapshot()` returns `None`.
2. **Adding `state` to that key is also wrong.** Two genuinely distinct `complete` snapshots can share a step index, where the later one carries newer history after an error rescue. Suppressing it leaves `latest_snapshot()` answering with stale history, which is worse than the duplicate the key was meant to prevent.

The key is a deterministic per-run sequence plus the step index and state, scoped by run id in the store. The sequence lives on the per-run instance `for_run` produces, so it starts at zero each run; replay re-executes the same saves in the same order and regenerates the key it already claimed, while a genuinely new save advances to one no earlier write has used. Events use the same shape, with `tool_call_id` in the key so repeated events within a step stay distinct.

Both wrong answers now have regression tests, so neither can be reintroduced quietly.

## Tests

DBOS on SQLite with `SetWorkflowID` forcing a replay, following the convention #707 merged. They prove a replayed run derives the same id rather than a second disconnected history, that timestamps come from the journaled write, that snapshots and tool effects are not duplicated, and that the two snapshot cases above both behave.

The pre-existing test asserting an eight-character run-id suffix encoded the behaviour this PR deliberately changes; it now asserts the whole caller-provided id survives and is deterministic across replay.

Step persistence 253 passed, repository suite 4,599 passed / 70 skipped, 100% statements and branches across the package, docs parity and snippets green, lint and pyright clean.

### CODE REVIEW

**Comment by DouweM** [comment]
One commit here (`072db7d9`) touches `pydantic_ai_harness/stackone/`, which has nothing to do with step persistence. Explaining why, since it looks out of place.

CI failed `lint` on this branch with:

```
pydantic_ai_harness/stackone/_capability.py:13:5 - error: "MCPToolsetClient" is not exported
from module "pydantic_ai_harness.stackone._toolset" (reportPrivateImportUsage)
```

on a file this PR does not touch, while `main` and the sibling PR #748 were green on the same base. It reproduced on a re-run, so it was not a flaky runner.

The cause: `_toolset.py` imports `MCPToolsetClient` inside the optional-import guard for the `stackone` extra, and `_capability.py` imports it from `_toolset` so that guard lives in one place, but it was never listed in `_toolset.__all__`. Pyright only reports that as a private import when the two modules land in *different* invocations, and the lint job runs pre-commit with `--all-files`, which batches by command length. Adding one test file anywhere in the tree shifts the batch boundary, which is why this PR surfaced a latent bug that #748 did not.

Naming it in `__all__` is what the diagnostic asks for and makes the result batch-independent. Happy to split it into its own PR if you would rather keep this one single-purpose, though it does block this branch as things stand.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 1]
```diff
@@ -182,30 +192,79 @@ async def for_run(self, ctx: RunContext[AgentDepsT]) -> AbstractCapability[Agent
...
     def _effective_run_id(self, ctx: RunContext[AgentDepsT]) -> str:
         if self.run_id is not None:
             return self.run_id
-        return ctx.run_id or str(uuid4())
+        if ctx.run_id is not None:
+            return ctx.run_id
+        raise RuntimeError('StepPersistence run id was not materialized by `for_run`.')
+
+    @durable_operation('register_run')
+    async def _register_run(self, record: RunRecord, reject_existing: bool) -> None:
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, and the same defect this PR set out to fix, missed for two of the four record types. `RunRecord.started_at` and `ToolEffectRecord.started_at` are both `field(default_factory=_utcnow)`, and both records were constructed in orchestration code before being handed to their durable operations, so the wall clock was read outside the durable unit exactly as it was for `StepEvent` and `ContinuableSnapshot`.

The tool-effect one is the more serious of the pair: it means every run reaching `before_tool_execute` fails inside a Temporal workflow, so the PR would have shipped claiming Temporal safety while breaking any run that calls a tool.

Fixed in `0124b4c1`: both records are now constructed inside their durable operations, so the recorded timestamp is journaled with the write. The DBOS replay test asserts both stay stable across a replay.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 2]
```diff
@@ -417,21 +555,19 @@ async def before_tool_execute(
         args: dict[str, Any],
     ) -> dict[str, Any]:
         run_id = self._effective_run_id(ctx)
-        await self.store.record_tool_effect(
+        await self._record_tool_effect(
             ToolEffectRecord(
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, and the same defect this PR set out to fix, missed for two of the four record types. `RunRecord.started_at` and `ToolEffectRecord.started_at` are both `field(default_factory=_utcnow)`, and both records were constructed in orchestration code before being handed to their durable operations, so the wall clock was read outside the durable unit exactly as it was for `StepEvent` and `ContinuableSnapshot`.

The tool-effect one is the more serious of the pair: it means every run reaching `before_tool_execute` fails inside a Temporal workflow, so the PR would have shipped claiming Temporal safety while breaking any run that calls a tool.

Fixed in `0124b4c1`: both records are now constructed inside their durable operations, so the recorded timestamp is journaled with the write. The DBOS replay test asserts both stay stable across a replay.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 3]
```diff
@@ -182,30 +192,79 @@ async def for_run(self, ctx: RunContext[AgentDepsT]) -> AbstractCapability[Agent
...
         if self.run_id is not None:
             return self.run_id
-        return ctx.run_id or str(uuid4())
+        if ctx.run_id is not None:
+            return ctx.run_id
+        raise RuntimeError('StepPersistence run id was not materialized by `for_run`.')
+
+    @durable_operation('register_run')
+    async def _register_run(self, record: RunRecord, reject_existing: bool) -> None:
+        if reject_existing and await self.store.get_run(run_id=record.run_id) is not None:
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, in both directions, and the four threads on this condition are one bug. `reject_existing=self.run_id is not None and self.run_id != ctx.run_id` cannot answer the question it is asked, because `for_run` always populates `self.run_id`, so an explicit id and a derived one are indistinguishable by comparison:

- a **derived** id differs from `ctx.run_id`, so `reject_existing` was `True` and a durable retry of `_register_run` raised `ValueError` after the store write had already succeeded, leaving recovery unable to complete;
- an **explicit** `StepPersistence(run_id="r")` used with `Agent.run(..., run_id="r")` compared equal, so `reject_existing` was `False` and the documented single-shot guarantee silently did not hold.

Fixed in `0124b4c1` by making registration identify itself rather than infer intent. `_registration_id` is its own durable operation returning a `uuid4()`, so the value is journaled: a replay of `before_run` receives the *recorded* id and `_register_run` recognises its own prior write as a retry, while a genuinely new run generates a new id and is rejected if it reuses an existing `run_id`. That is the standard way to make a nondeterministic value replay-safe, and it removes the guesswork the comparison was doing.

Covered by three tests: durable replay, same-claim retry, and explicit reuse.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 4]
```diff
@@ -246,22 +393,17 @@ async def before_run(self, ctx: RunContext[AgentDepsT]) -> None:
...
+        await self._register_run(
             RunRecord(
                 run_id=run_id,
                 conversation_id=ctx.conversation_id,
                 parent_run_id=self.parent_run_id,
                 agent_name=self.agent_name,
                 metadata=dict(self.metadata),
-            )
+            ),
+            reject_existing=self.run_id is not None and self.run_id != ctx.run_id,
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, in both directions, and the four threads on this condition are one bug. `reject_existing=self.run_id is not None and self.run_id != ctx.run_id` cannot answer the question it is asked, because `for_run` always populates `self.run_id`, so an explicit id and a derived one are indistinguishable by comparison:

- a **derived** id differs from `ctx.run_id`, so `reject_existing` was `True` and a durable retry of `_register_run` raised `ValueError` after the store write had already succeeded, leaving recovery unable to complete;
- an **explicit** `StepPersistence(run_id="r")` used with `Agent.run(..., run_id="r")` compared equal, so `reject_existing` was `False` and the documented single-shot guarantee silently did not hold.

Fixed in `0124b4c1` by making registration identify itself rather than infer intent. `_registration_id` is its own durable operation returning a `uuid4()`, so the value is journaled: a replay of `before_run` receives the *recorded* id and `_register_run` recognises its own prior write as a retry, while a genuinely new run generates a new id and is rejected if it reuses an existing `run_id`. That is the standard way to make a nondeterministic value replay-safe, and it removes the guesswork the comparison was doing.

Covered by three tests: durable replay, same-claim retry, and explicit reuse.

[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, in both directions, and the four threads on this condition are one bug. `reject_existing=self.run_id is not None and self.run_id != ctx.run_id` cannot answer the question it is asked, because `for_run` always populates `self.run_id`, so an explicit id and a derived one are indistinguishable by comparison:

- a **derived** id differs from `ctx.run_id`, so `reject_existing` was `True` and a durable retry of `_register_run` raised `ValueError` after the store write had already succeeded, leaving recovery unable to complete;
- an **explicit** `StepPersistence(run_id="r")` used with `Agent.run(..., run_id="r")` compared equal, so `reject_existing` was `False` and the documented single-shot guarantee silently did not hold.

Fixed in `0124b4c1` by making registration identify itself rather than infer intent. `_registration_id` is its own durable operation returning a `uuid4()`, so the value is journaled: a replay of `before_run` receives the *recorded* id and `_register_run` recognises its own prior write as a retry, while a genuinely new run generates a new id and is rejected if it reuses an existing `run_id`. That is the standard way to make a nondeterministic value replay-safe, and it removes the guesswork the comparison was doing.

Covered by three tests: durable replay, same-claim retry, and explicit reuse.

[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, in both directions, and the four threads on this condition are one bug. `reject_existing=self.run_id is not None and self.run_id != ctx.run_id` cannot answer the question it is asked, because `for_run` always populates `self.run_id`, so an explicit id and a derived one are indistinguishable by comparison:

- a **derived** id differs from `ctx.run_id`, so `reject_existing` was `True` and a durable retry of `_register_run` raised `ValueError` after the store write had already succeeded, leaving recovery unable to complete;
- an **explicit** `StepPersistence(run_id="r")` used with `Agent.run(..., run_id="r")` compared equal, so `reject_existing` was `False` and the documented single-shot guarantee silently did not hold.

Fixed in `0124b4c1` by making registration identify itself rather than infer intent. `_registration_id` is its own durable operation returning a `uuid4()`, so the value is journaled: a replay of `before_run` receives the *recorded* id and `_register_run` recognises its own prior write as a retry, while a genuinely new run generates a new id and is rejected if it reuses an existing `run_id`. That is the standard way to make a nondeterministic value replay-safe, and it removes the guesswork the comparison was doing.

Covered by three tests: durable replay, same-claim retry, and explicit reuse.

[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, in both directions, and the four threads on this condition are one bug. `reject_existing=self.run_id is not None and self.run_id != ctx.run_id` cannot answer the question it is asked, because `for_run` always populates `self.run_id`, so an explicit id and a derived one are indistinguishable by comparison:

- a **derived** id differs from `ctx.run_id`, so `reject_existing` was `True` and a durable retry of `_register_run` raised `ValueError` after the store write had already succeeded, leaving recovery unable to complete;
- an **explicit** `StepPersistence(run_id="r")` used with `Agent.run(..., run_id="r")` compared equal, so `reject_existing` was `False` and the documented single-shot guarantee silently did not hold.

Fixed in `0124b4c1` by making registration identify itself rather than infer intent. `_registration_id` is its own durable operation returning a `uuid4()`, so the value is journaled: a replay of `before_run` receives the *recorded* id and `_register_run` recognises its own prior write as a retry, while a genuinely new run generates a new id and is rejected if it reuses an existing `run_id`. That is the standard way to make a nondeterministic value replay-safe, and it removes the guesswork the comparison was doing.

Covered by three tests: durable replay, same-claim retry, and explicit reuse.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 5]
```diff
@@ -182,30 +192,79 @@ async def for_run(self, ctx: RunContext[AgentDepsT]) -> AbstractCapability[Agent
         """
         inferred_parent = self.parent_run_id if self.parent_run_id is not None else current_run_id.get()
         resolved_run_id = self.run_id or self._derive_run_id(ctx)
-        if resolved_run_id == self.run_id and inferred_parent == self.parent_run_id:
-            return self
         return replace(self, run_id=resolved_run_id, parent_run_id=inferred_parent)
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py:196]: Real, and a direct consequence of this PR switching to the whole `ctx.run_id`. The previous code truncated to eight hex characters and so was path-safe by accident; `FileStepStore` validates against `^[A-Za-z0-9_.-]{1,200}$`, and a caller can pass `Agent.run(..., run_id="tenant/run")`.

Fixed in `0124b4c1`. Using the whole run id stays (a caller-supplied run id is not a UUID, so a suffix of it is arbitrary and collision-prone), so the derived value is normalized instead: the `(agent_name, ctx.run_id)` pair is length-delimited and base64url encoded, which is injective and lands inside the store's alphabet. Values exceeding the 200-character limit raise rather than truncating into a collision. Tested with `tenant/...` ids, distinct-input injectivity, file-store registration, and overflow.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 6]
```diff
@@ -182,30 +192,79 @@ async def for_run(self, ctx: RunContext[AgentDepsT]) -> AbstractCapability[Agent
...
         kind: EventKind,
         tool_call_id: str | None = None,
         tool_name: str | None = None,
         error: str | None = None,
-    ) -> StepEvent:
-        return StepEvent(
+    ) -> None:
+        event_index = self._event_sequence
+        self._event_sequence += 1
+        await self._append_event(
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py:299]: Real, and my fault in scoping the work: the brief for this PR named only the in-memory and Mongo stores, so `FileStepStore` and `SqliteStepStore` kept appending regardless of the key. That made the `idempotency_key` on `StepEvent` and `ContinuableSnapshot` decorative for two of the four backends.

Fixed in `0124b4c1`: both now persist and suppress the key. Tests replay keyed writes against each and confirm unkeyed writes stay append-only, so nothing changes for a caller constructing records directly.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 7]
```diff
@@ -216,6 +275,94 @@ def _make_event(
...
+                agent_name=agent_name,
+                timestamp=datetime.now(timezone.utc),
+                state=state,
+                # The store scopes keys by run id. Within that scope the sequence distinguishes
+                # separate save calls, step_index identifies the graph position, and state keeps
+                # settled and interrupted histories distinct. `for_run` creates the capability
+                # instance that owns this zero-based counter, so each run starts at zero. Replay
+                # executes the same saves in order and regenerates the claimed key; a new save
+                # advances to a key no earlier write has used.
+                idempotency_key=f'{snapshot_index}:{step_index}:{state}',
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py:342]: Real, and my fault in scoping the work: the brief for this PR named only the in-memory and Mongo stores, so `FileStepStore` and `SqliteStepStore` kept appending regardless of the key. That made the `idempotency_key` on `StepEvent` and `ContinuableSnapshot` decorative for two of the four backends.

Fixed in `0124b4c1`: both now persist and suppress the key. Tests replay keyed writes against each and confirm unkeyed writes stay append-only, so nothing changes for a caller constructing records directly.

File: pydantic_ai_harness/step_persistence/_mongo.py [hunk 8]
```diff
@@ -149,7 +159,9 @@ async def _next_seq(self, name: str) -> int:
 
     async def register_run(self, record: RunRecord) -> None:
         await self._ensure_indexes()
-        await self._db['runs'].insert_one({'_id': record.run_id, **_run_to_dict(record)})
+        await self._db['runs'].replace_one(
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_mongo.py]: Real. `replace_one(..., upsert=True)` made the separate duplicate check racy, so two concurrent runs claiming the same explicit `run_id` overwrote each other rather than the second failing.

Fixed in `0124b4c1`: registration is an atomic `insert_one`, and `DuplicateKeyError` is translated into the same `ValueError` the non-racing path raises, so callers see one behaviour regardless of which check caught it.

File: pydantic_ai_harness/step_persistence/_store.py [hunk 9]
```diff
@@ -214,13 +214,23 @@ async def list_runs(
...
 
     async def list_events(self, *, run_id: str) -> list[StepEvent]:
         return list(self._events.get(run_id, ()))
 
     async def save_snapshot(self, snapshot: ContinuableSnapshot) -> None:
         snaps = self._snapshots[snapshot.run_id]
+        if snapshot.idempotency_key is not None:
+            # Applied keys live only beside their run's records, so deleting a run also removes
+            # the suppression state and no independent retention policy is needed.
+            if any(prior.idempotency_key == snapshot.idempotency_key for prior in snaps):
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_store.py:251]: Real, and a good catch on an interaction I would not have found: tying suppression state to retained snapshots means pruning discards the evidence that a key was already applied, so with `max_snapshots_per_run=1` a replayed write of a pruned snapshot inserts as the newest and evicts the newer one, leaving `latest_snapshot()` stale.

Fixed in `0124b4c1` by keeping suppression state independent of snapshot retention in every store. The in-memory footprint is bounded by one high-water sequence plus the current keys per run rather than growing with history, and the comment says so. Tests replay a pruned older snapshot and assert the newer one is still latest.

File: pydantic_ai_harness/step_persistence/_capability.py [hunk 10]
```diff
@@ -81,14 +86,18 @@ class StepPersistence(AbstractCapability[AgentDepsT]):
     `Agent.run(..., message_history=...)`.
     """
 
+    # Override the inherited default ID because durable-operation recovery needs a stable identity.
+    _: KW_ONLY
+    id: str | None = 'step_persistence'
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_capability.py]: Real, and the worst kind of irony: the mechanism added to prevent a backward-compatibility break introduced one. `_: KW_ONLY` sat at the top of the field list, so `store`, `agent_name`, `run_id` and everything after them became keyword-only, and `StepPersistence(InMemoryStepStore())` raised `TypeError`.

Fixed in `a557c141` with your suggested form, `field(default="step_persistence", kw_only=True)`, rather than by relocating the marker. Both work, but this one does not depend on where the field sits, so a later field reordering cannot silently re-break positional construction. `test_store_remains_positional` pins it.

For anyone auditing the sibling PRs: #748 places the marker after all its public fields, so `SpendLimits` only made `id` and its `init=False` privates keyword-only and was never affected.

File: pydantic_ai_harness/step_persistence/_store.py [hunk 11]
```diff
@@ -575,9 +631,13 @@ def _sync_save_snapshot(self, snapshot: ContinuableSnapshot, messages_json: obje
             'timestamp': snapshot.timestamp.isoformat(),
             'state': snapshot.state,
             'messages': messages_json,
+            'idempotency_key': snapshot.idempotency_key,
         }
         seq = self._next_snapshot_seq(snap_dir)
         _atomic_write_text(snap_dir / f'{seq}.json', json.dumps(payload))
+        if snapshot.idempotency_key is not None:
+            with key_path.open('a', encoding='utf-8') as fp:
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_harness/step_persistence/_store.py:645]: Real. Keys were written raw one per line and read back with `splitlines()`, so a key containing a separator was recorded as two keys and a later distinct key equal to either half was treated as already-applied and silently dropped, losing a real snapshot write.

Fixed in `a557c141` by JSON-encoding each key on write and decoding with typed validation on read, rather than by rejecting separators. Encoding is the safer of the two here because `splitlines()` splits on nine characters, not just `\n` (`\r`, `\x0b`, `\x0c`, `\x1c`, `\x1d`, `\x1e`, `\x85`, ` `, ` `), so a validation-based fix has nine chances to miss one.

Event keys needed no change: events are already stored as complete JSON records rather than bare lines. `test_snapshot_key_line_separator_does_not_suppress_distinct_key` proves the case you described.
